conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureAsyncRequestProcessing.java

```diff
@@ -116,8 +116,15 @@
             ListenableFuture<T> future, ReturnValueWriter<T> returnValueWriter, HttpServerExchange exchange) {
         // Attach data to the exchange in order to reuse a stateless listener
         exchange.putAttachment(FUTURE, future);
-        // Cancel the future if the exchange is completed before the future is done
-        exchange.addExchangeCompleteListener(COMPLETION_LISTENER);
+        // If the exchange is complete we still follow this path to provide standard logging.
+        if (exchange.isComplete()) {
```

Comment:
Is this not racy in any way? Is it not possible that exchange would complete in between you calling exchange.isComplete and exchange.addExchangeCompleteListener?

Reply:
It shouldn't be since the data both read from is non-volatile.

Reply:
Sweet, I see that now.